File: src/main/java/org/moinex/ui/dialog/wallettransaction/TransferController.java
package org.moinex.ui.dialog.wallettransaction;

import jakarta.persistence.EntityNotFoundException;
import java.text.MessageFormat;
import java.time.LocalDate;
import java.util.List;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.collections.transformation.FilteredList;
import javafx.fxml.FXML;
import javafx.geometry.Pos;
import javafx.scene.control.DatePicker;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.wallettransaction.Transfer;
import org.moinex.service.I18nService;
import org.moinex.service.WalletTransactionService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Manage Transfers dialog
 */
@Controller
@NoArgsConstructor
public class TransferController {
    @FXML private TableView<Transfer> transfersTableView;
    @FXML private TextField searchField;
    @FXML private DatePicker transfersEndDatePicker;
    @FXML private DatePicker transfersStartDatePicker;

    private ConfigurableApplicationContext springContext;
    private WalletTransactionService walletTransactionService;
    private I18nService i18nService;

    private final ObservableList<Transfer> masterData = FXCollections.observableArrayList();
    private FilteredList<Transfer> filteredData;

    @Autowired
    public TransferController(
            WalletTransactionService walletTransactionService,
            ConfigurableApplicationContext springContext,
            I18nService i18nService) {
        this.walletTransactionService = walletTransactionService;
        this.springContext = springContext;
        this.i18nService = i18nService;
    }

    @FXML
    public void initialize() {
        loadTransfersFromDatabase();

        configureTableView();
        configureDatePickers();
        setupFilterListeners();

        filteredData = new FilteredList<>(masterData, p -> true);
        updateTableView();
    }

    /**
     * Configures the TableView columns to display Transfer data.
     */
    private void configureTableView() {
        TableColumn<Transfer, Integer> idColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.WALLETTRANSACTION_TABLE_ID));
        idColumn.setCellValueFactory(param -> new SimpleObjectProperty<>(param.getValue().getId()));

        TableColumn<Transfer, String> dateColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.WALLETTRANSACTION_TABLE_DATE));
        dateColumn.setCellValueFactory(
                param ->
                        new SimpleStringProperty(
                                UIUtils.formatDateForDisplay(
                                        param.getValue().getDate(), i18nService)));

        TableColumn<Transfer, String> descriptionColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys.WALLETTRANSACTION_TABLE_DESCRIPTION));
        descriptionColumn.setCellValueFactory(
                param -> new SimpleStringProperty(param.getValue().getDescription()));

        TableColumn<Transfer, String> amountColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.WALLETTRANSACTION_TABLE_AMOUNT));
        amountColumn.setCellValueFactory(
                param ->
                        new SimpleStringProperty(
                                UIUtils.formatCurrency(param.getValue().getAmount())));

        TableColumn<Transfer, String> senderColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.WALLETTRANSACTION_TABLE_SENDER));
        senderColumn.setCellValueFactory(
                param -> new SimpleStringProperty(param.getValue().getSenderWallet().getName()));

        TableColumn<Transfer, String> receiverColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.WALLETTRANSACTION_TABLE_RECEIVER));
        receiverColumn.setCellValueFactory(
                param -> new SimpleStringProperty(param.getValue().getReceiverWallet().getName()));

        TableColumn<Transfer, String> categoryColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.WALLETTRANSACTION_TABLE_CATEGORY));
        categoryColumn.setCellValueFactory(
                param -> {
                    if (param.getValue().getCategory() != null) {
                        return new SimpleStringProperty(param.getValue().getCategory().getName());
                    }
                    return new SimpleStringProperty("-");
                });

        UIUtils.alignTableColumn(
                List.of(
                        idColumn,
                        dateColumn,
                        amountColumn,
                        senderColumn,
                        receiverColumn,
                        categoryColumn),
                Pos.CENTER);

        transfersTableView.getColumns().add(idColumn);
        transfersTableView.getColumns().add(dateColumn);
        transfersTableView.getColumns().add(descriptionColumn);
        transfersTableView.getColumns().add(amountColumn);
        transfersTableView.getColumns().add(senderColumn);
        transfersTableView.getColumns().add(receiverColumn);
        transfersTableView.getColumns().add(categoryColumn);
    }

    /**
     * Configures the date pickers with appropriate formatters and default values.
     */
    private void configureDatePickers() {
        UIUtils.setDatePickerFormat(transfersStartDatePicker, i18nService);
        UIUtils.setDatePickerFormat(transfersEndDatePicker, i18nService);

        transfersStartDatePicker.setValue(LocalDate.now().withDayOfMonth(1));
        transfersEndDatePicker.setValue(LocalDate.now().plusMonths(1).withDayOfMonth(1));
    }

    /**
     * Loads all transfers from the database into the master list and sets up the filtered list.
     */
    private void loadTransfersFromDatabase() {
        masterData.setAll(walletTransactionService.getAllTransfers());
    }

    /**
     * Sets up listeners on the search field and date pickers to filter the table view.
     */
    private void setupFilterListeners() {
        searchField.textProperty().addListener((obs, old, val) -> updateTableView());
        transfersStartDatePicker.valueProperty().addListener((obs, old, val) -> updateTableView());
        transfersEndDatePicker.valueProperty().addListener((obs, old, val) -> updateTableView());
    }

    /**
     * Applies the current filters from the UI to the table view data.
     */
    private void updateTableView() {
        String searchText = searchField.getText().toLowerCase();
        LocalDate startDate = transfersStartDatePicker.getValue();
        LocalDate endDate = transfersEndDatePicker.getValue();

        filteredData.setPredicate(
                transfer -> {
                    boolean matchesSearch =
                            searchText.isEmpty()
                                    || transfer.getDescription().toLowerCase().contains(searchText)
                                    || transfer.getSenderWallet()
                                            .getName()
                                            .toLowerCase()
                                            .contains(searchText)
                                    || transfer.getReceiverWallet()
                                            .getName()
                                            .toLowerCase()
                                            .contains(searchText)
                                    || String.valueOf(transfer.getId()).contains(searchText);

                    boolean matchesDate = true;
                    LocalDate transferDate = transfer.getDate().toLocalDate();
                    if (startDate != null && transferDate.isBefore(startDate)) {
                        matchesDate = false;
                    }
                    if (endDate != null && transferDate.isAfter(endDate)) {
                        matchesDate = false;
                    }

                    return matchesSearch && matchesDate;
                });

        transfersTableView.setItems(filteredData);
    }

    @FXML
    private void handleAdd() {
        WindowUtils.openModalWindow(
                Constants.ADD_TRANSFER_FXML,
                i18nService.tr(
                        Constants.TranslationKeys.WALLETTRANSACTION_DIALOG_ADD_NEW_TRANSFER_TITLE),
                springContext,
                (AddTransferController controller) -> {},
                List.of(
                        () -> {
                            loadTransfersFromDatabase();
                            updateTableView();
                        }));
    }

    @FXML
    private void handleEdit() {
        Transfer selectedTransfer = transfersTableView.getSelectionModel().getSelectedItem();
        if (selectedTransfer == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_NO_TRANSFER_SELECTED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_NO_TRANSFER_SELECTED_EDIT_MESSAGE));
            return;
        }

        WindowUtils.openModalWindow(
                Constants.EDIT_TRANSFER_FXML,
                i18nService.tr(
                        Constants.TranslationKeys.WALLETTRANSACTION_DIALOG_EDIT_TRANSFER_TITLE),
                springContext,
                (EditTransferController controller) -> controller.setTransfer(selectedTransfer),
                List.of(this::loadTransfersFromDatabase));
    }

    @FXML
    private void handleDelete() {
        Transfer selectedTransfer = transfersTableView.getSelectionModel().getSelectedItem();
        if (selectedTransfer == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_NO_TRANSFER_SELECTED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_NO_TRANSFER_SELECTED_DELETE_MESSAGE));
            return;
        }

        if (WindowUtils.showConfirmationDialog(
                i18nService.tr(
                        Constants.TranslationKeys.WALLETTRANSACTION_DIALOG_DELETE_TRANSFER_TITLE),
                MessageFormat.format(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .WALLETTRANSACTION_DIALOG_DELETE_TRANSFER_MESSAGE),
                        selectedTransfer.getId()),
                i18nService.getBundle())) {
            try {
                walletTransactionService.deleteTransfer(selectedTransfer.getId());
                WindowUtils.showSuccessDialog(
                        i18nService.tr(
                                Constants.TranslationKeys.WALLETTRANSACTION_DIALOG_SUCCESS_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .WALLETTRANSACTION_DIALOG_TRANSFER_DELETED_MESSAGE));

                loadTransfersFromDatabase();
                updateTableView();

            } catch (EntityNotFoundException e) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(
                                Constants.TranslationKeys.WALLETTRANSACTION_DIALOG_ERROR_TITLE),
                        e.getMessage());
            }
        }
    }

    @FXML
    private void handleCancel() {
        Stage stage = (Stage) searchField.getScene().getWindow();
        stage.close();
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/wallettransaction/view/TransferController.kt
/*
 * Filename: TransferController.kt (original filename: TransferController.java)
 * Created on: October  4, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 21/03/2026
 */

package org.moinex.ui.dialog.wallettransaction.view

import jakarta.persistence.EntityNotFoundException
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.collections.FXCollections
import javafx.collections.transformation.FilteredList
import javafx.fxml.FXML
import javafx.geometry.Pos
import javafx.scene.control.DatePicker
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constant.Files
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.wallettransaction.Transfer
import org.moinex.service.PreferencesService
import org.moinex.service.wallet.WalletService
import org.moinex.ui.dialog.wallettransaction.create.AddTransferController
import org.moinex.ui.dialog.wallettransaction.update.EditTransferController
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.stereotype.Controller
import java.text.MessageFormat
import java.time.LocalDate

@Controller
class TransferController(
    private val walletService: WalletService,
    private val springContext: ConfigurableApplicationContext,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var transfersTableView: TableView<Transfer>

    @FXML
    private lateinit var searchField: TextField

    @FXML
    private lateinit var transfersEndDatePicker: DatePicker

    @FXML
    private lateinit var transfersStartDatePicker: DatePicker

    private val masterData = FXCollections.observableArrayList<Transfer>()
    private lateinit var filteredData: FilteredList<Transfer>

    @FXML
    fun initialize() {
        loadTransfersFromDatabase()
        configureTableView()
        configureDatePickers()
        setupFilterListeners()

        filteredData = FilteredList(masterData) { true }
        updateTableView()
    }

    @FXML
    private fun handleAdd() {
        WindowUtils.openModalWindow(
            Files.ADD_TRANSFER_FXML,
            preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_ADD_NEW_TRANSFER_TITLE),
            springContext,
            { _: AddTransferController -> },
            listOf(
                Runnable {
                    loadTransfersFromDatabase()
                    updateTableView()
                },
            ),
        )
    }

    @FXML
    private fun handleEdit() {
        val selectedTransfer = transfersTableView.selectionModel.selectedItem
        if (selectedTransfer == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_NO_TRANSFER_SELECTED_TITLE),
                preferencesService.translate(
                    TranslationKeys.WALLETTRANSACTION_DIALOG_NO_TRANSFER_SELECTED_EDIT_MESSAGE,
                ),
            )
            return
        }

        WindowUtils.openModalWindow(
            Files.EDIT_TRANSFER_FXML,
            preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EDIT_TRANSFER_TITLE),
            springContext,
            { controller: EditTransferController -> controller.setTransfer(selectedTransfer) },
            listOf(Runnable { loadTransfersFromDatabase() }),
        )
    }

    @FXML
    private fun handleDelete() {
        val selectedTransfer = transfersTableView.selectionModel.selectedItem
        if (selectedTransfer == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_NO_TRANSFER_SELECTED_TITLE),
                preferencesService.translate(
                    TranslationKeys.WALLETTRANSACTION_DIALOG_NO_TRANSFER_SELECTED_DELETE_MESSAGE,
                ),
            )
            return
        }

        if (WindowUtils.showConfirmationDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_DELETE_TRANSFER_TITLE),
                MessageFormat.format(
                    preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_DELETE_TRANSFER_MESSAGE),
                    selectedTransfer.id,
                ),
                preferencesService.bundle,
            )
        ) {
            runCatching {
                walletService.deleteTransfer(selectedTransfer.id!!)
                loadTransfersFromDatabase()
                updateTableView()
            }.onFailure { e ->
                if (e is EntityNotFoundException) {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_ERROR_TITLE),
                        e.message ?: "Unknown error",
                    )
                } else {
                    throw e
                }
            }
        }
    }

    @FXML
    private fun handleCancel() {
        val stage = searchField.scene.window as Stage
        stage.close()
    }

    private fun configureDatePickers() {
        UIUtils.setDatePickerFormat(transfersStartDatePicker)
        UIUtils.setDatePickerFormat(transfersEndDatePicker)

        transfersStartDatePicker.value = LocalDate.now().withDayOfMonth(1)
        transfersEndDatePicker.value = LocalDate.now().plusMonths(1).withDayOfMonth(1)
    }

    private fun loadTransfersFromDatabase() {
        masterData.setAll(walletService.getAllTransfers())
    }

    private fun setupFilterListeners() {
        searchField.textProperty().addListener { _, _, _ -> updateTableView() }
        transfersStartDatePicker.valueProperty().addListener { _, _, _ -> updateTableView() }
        transfersEndDatePicker.valueProperty().addListener { _, _, _ -> updateTableView() }
    }

    private fun updateTableView() {
        val searchText = searchField.text.lowercase()
        val startDate = transfersStartDatePicker.value
        val endDate = transfersEndDatePicker.value

        filteredData.setPredicate { transfer ->
            val matchesSearch =
                searchText.isEmpty() ||
                    transfer.description?.lowercase()?.contains(searchText) ?: true ||
                    transfer.senderWallet.name
                        .lowercase()
                        .contains(searchText) ||
                    transfer.receiverWallet.name
                        .lowercase()
                        .contains(searchText) ||
                    transfer.id.toString().contains(searchText)

            var matchesDate = true
            val transferDate = transfer.date.toLocalDate()
            if (startDate != null && transferDate.isBefore(startDate)) {
                matchesDate = false
            }
            if (endDate != null && transferDate.isAfter(endDate)) {
                matchesDate = false
            }

            matchesSearch && matchesDate
        }

        transfersTableView.items = filteredData
    }

    private fun configureTableView() {
        val idColumn =
            TableColumn<Transfer, Int>(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_TABLE_ID),
            ).apply {
                setCellValueFactory { SimpleObjectProperty(it.value.id) }
                UIUtils.alignTableColumn(this, Pos.CENTER)
            }

        val dateColumn =
            TableColumn<Transfer, String>(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_TABLE_DATE),
            ).apply {
                setCellValueFactory {
                    SimpleStringProperty(UIUtils.formatDateForDisplay(it.value.date))
                }
                UIUtils.alignTableColumn(this, Pos.CENTER)
            }

        val descriptionColumn =
            TableColumn<Transfer, String>(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_TABLE_DESCRIPTION),
            ).apply {
                setCellValueFactory { SimpleStringProperty(it.value.description) }
            }

        val amountColumn =
            TableColumn<Transfer, String>(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_TABLE_AMOUNT),
            ).apply {
                setCellValueFactory {
                    SimpleStringProperty(UIUtils.formatCurrency(it.value.amount))
                }
                UIUtils.alignTableColumn(this, Pos.CENTER)
            }

        val senderColumn =
            TableColumn<Transfer, String>(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_TABLE_SENDER),
            ).apply {
                setCellValueFactory { SimpleStringProperty(it.value.senderWallet.name) }
                UIUtils.alignTableColumn(this, Pos.CENTER)
            }

        val receiverColumn =
            TableColumn<Transfer, String>(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_TABLE_RECEIVER),
            ).apply {
                setCellValueFactory { SimpleStringProperty(it.value.receiverWallet.name) }
                UIUtils.alignTableColumn(this, Pos.CENTER)
            }

        val categoryColumn =
            TableColumn<Transfer, String>(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_TABLE_CATEGORY),
            ).apply {
                setCellValueFactory {
                    SimpleStringProperty(it.value.category?.name ?: "-")
                }
                UIUtils.alignTableColumn(this, Pos.CENTER)
            }

        transfersTableView.columns.addAll(
            idColumn,
            dateColumn,
            descriptionColumn,
            amountColumn,
            senderColumn,
            receiverColumn,
            categoryColumn,
        )
    }
}
